### PR TITLE
implement simple NTP service to check if local time is valid

### PR DESCRIPTION
**Pull request by vogel**
Signed-off-by: Rafał Przemysław Malinowski <rafal@stellar.org>

### CODE REVIEW

**Comment by jedmccaleb** [comment]
Cool. In the future you should divide PRs into more commits if there are things that can be analyzed separately. For example this could be at least two commits; one for the IssueManager and one for the NTP stuff. 
I think IssueManager should be refactored a bit. Ideally it would replace this "extra" thing we have now. So I think the interface we want is something like `IssueManager::addIssue(int ID, string displayedText)`  and `IssueManager::removeIssue(int ID)`   
This way we can include dynamic info in the displayed issues.

File: src/main/CommandHandler.cpp [hunk 1]
```diff
@@ -437,7 +438,7 @@ CommandHandler::info(std::string const& params, std::string& retStr)
 {
     Json::Value root;
 
-    LedgerManager& lm = mApp.getLedgerManager();
+    auto& lm = mApp.getLedgerManager();
```
Comment (jedmccaleb) [inline src/main/CommandHandler.cpp]: why did you change to auto?

File: src/util/NtpWork.h [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+     * within acceptable tolerance (passed in constructor).
+     * If work is unfinished or unsuccessfull, behavior of this method is undefined.
+     */
+    bool isSynchronized() const;
+
+private:
+    std::string mNtpServer;
+    std::chrono::seconds mTolerance;
+    VirtualTimer mTimeoutTimer;
+    bool mIsSynchronized;
```
Comment (jedmccaleb) [inline src/util/NtpWork.h]: not initialized

File: src/main/ApplicationImpl.cpp [hunk 3]
```diff
@@ -198,6 +206,10 @@ ApplicationImpl::getNetworkID() const
 ApplicationImpl::~ApplicationImpl()
 {
     LOG(INFO) << "Application destructing";
+    if (mNtpSynchronizationChecker)
+    {
+        mNtpSynchronizationChecker->shutdown();
```
Comment (jedmccaleb) [inline src/main/ApplicationImpl.cpp]: Does this need to be called in the destructor? It is also called in gracefulStop   seems like it should just be called in one or the other

File: src/main/NtpSynchronizationChecker.cpp [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+            return;
+        }
+        self->start();
+    }, &VirtualTimer::onFailureNoop);
+}
+
+void
+NtpSynchronizationChecker::start()
+{
+    if (mIsShutdown)
```
Comment (jedmccaleb) [inline src/main/NtpSynchronizationChecker.cpp]: How can this happen?

Comment (vogel) [inline src/main/NtpSynchronizationChecker.cpp]: start is called from scheduleNextCheck

File: src/main/ApplicationImpl.cpp (hunk 3 above)
Comment (vogel) [inline src/main/ApplicationImpl.cpp]: I was just following places where mProcessManager->shutdown() is called, as
mNtpSynchronizationChecker is similar to mProcessManager. Checking if this is really needed would require separate investigation.

File: src/util/NtpWork.h (hunk 2 above)
Comment (vogel) [inline src/util/NtpWork.h]: Will initialize to true.

File: src/main/CommandHandler.cpp (hunk 1 above)
Comment (vogel) [inline src/main/CommandHandler.cpp]: Without real reason, I was just changing code around it. I'll revert it back and rebase so this commit will not have this random change.

**Comment by vogel** [comment]
Currently "extra" contains last history entry, so I'm not exactly sure if it matches the name 'issue'.
How would you like to have the information presented in the "issues" field?

Maybe like this:

"issues": {
  "history": "content of current-extra-field",
  "ntp": "Local time is not synchronized with NTP time.",
  "other-category": "other information"
}

?

**Comment by jedmccaleb** [comment]
Yeah maybe "issues" isn't the best name for it. I think maybe just:
"extra" or "status": {
"content of current-extra-field",
"Local time is not synchronized with NTP time.",
"other information"
}
doesn't seem like we need the title of the field. It should be clear from the displayed string

**Comment by vogel** [comment]
Fixed travis build
